lib/publiccloud/instance.pm

```diff
@@ -91,4 +96,111 @@ sub upload_log {
     assert_script_run("test -d '$tmpdir' && rm -rf '$tmpdir'");
 }
 
+=head2 wait_fo_guestregister
+
+    wait_for_guestregister([timeout => 300]);
+
+Run command C<systemctl is-active guestregister> on the instance in a loop
+for max C<timeout> seconds until it will return inactive.
+=cut
+sub wait_for_guestregister
+{
+    my ($self, %args) = @_;
+    $args{timeout} //= 300;
+    my $start_time = time();
+    my $last_info  = 0;
+
+    while (time() - $start_time < $args{timeout}) {
+        my $out = $self->run_ssh_command(cmd => 'sudo systemctl is-active guestregister', proceed_on_failure => 1, quiet => 1);
+        if ($out eq 'inactive') {
+            return time() - $start_time;
+        }
+        if (time() - $last_info > 10) {
+            record_info('WAIT', 'Wait for guest register: ' . $out);
+            $last_info = time();
+        }
+        sleep 1;
+    }
+    die('guestregister didn\'t end in expected timeout=' . $args{timeout});
+}
+
+=head2 check_ssh_port
+
+    check_ssh_port([timeout => 600] [, proceed_on_failure => 0])
+
+Check if the SSH port of the instance is reachable and open.
+=cut
+sub check_ssh_port
+{
+    my ($self, %args) = @_;
+    $args{timeout}            //= 600;
+    $args{proceed_on_failure} //= 0;
+    my $start_time = time();
+
+    while ((my $duration = time() - $start_time) < $args{timeout}) {
+        return $duration if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
+        sleep 1;
+    }
+    croak(sprintf("Unable to reach SSH port of instance %s with public IP:%s within %d seconds",
+            $self->{instance_id}, $self->{public_ip}, $self->{timeout}))
+      unless ($args{proceed_on_failure});
+    return;
+}
+
+=head2 softreboot
+
+    ($shutdown_time, $bootup_time) = softreboot([timeout => 600]);
+
+Does a softreboot of the instance by running the command C<shutdown -r>.
+Return an array of two values, first one is the time till the instance isn't
+reachable anymore. The second one is the estimated bootup time.
+=cut
+sub softreboot
+{
+    my ($self, %args) = @_;
+    $args{timeout} //= 600;
+
+    my $duration;
+
+    $self->run_ssh_command(cmd => 'sudo shutdown -r +1');
+    # skip the one minute waiting
+    sleep 60;
```

Comment:
why not just increase `args{timeout}` in case we need wait more ? find it confusing that we have two ways to increase waiting time ...

Reply:
The `shutdown -r +1` do the shutdown in 1 minute. And I just wait this time, as this doesn't belong to the `shutdown-time`.
If I remove the `+1` and run `shutdown -r now`, I run in trouble with `run_ssh_command()` as the connection close and the return value isn't `0`.

Reply:
you can just remove this sleep and increase `$args{timeout}` by 60 ?